compiler_gym/leaderboard/llvm_instcount.py

```diff
@@ -84,61 +87,69 @@
+        state = env.state.copy()
+        state.walltime = timer.time
+        env.close()
+        return benchmark, state
+    except:
+        # If the code fails, then return None and we restart the same job again.
+        return benchmark, None
+
+class _EvalPolicyWorker(mp.Process):
     """Worker thread to evaluate a policy."""
 
     def __init__(
         self,
-        env: LlvmEnv,
+        num_process: int,
         benchmarks: List[str],
         policy: Policy,
-        init_states: List[CompilerEnvState],
+        q: mp.Queue
     ):
         super().__init__()
-        self.env = env
+        self.num_process = num_process
         self.benchmarks = benchmarks
         self.policy = policy
-        self.states: List[CompilerEnvState] = init_states
-        self.alive = True
+        self.q = q
 
     def run(self):
-        # Determine if we need to print a header.
-        header = (
-            not Path(FLAGS.leaderboard_results).is_file()
-            or os.stat(FLAGS.leaderboard_results).st_size == 0
-        )
-        with CompilerEnvStateWriter(
-            open(FLAGS.leaderboard_results, "a"), header=header
-        ) as writer:
-            for benchmark in self.benchmarks:
-                self.env.reset(benchmark=benchmark)
-                with Timer() as timer:
-                    self.policy(self.env)
-
-                # Sanity check that the policy didn't change the expected
-                # experimental setup.
-                assert self.env.in_episode, "Environment is no longer in an episode"
-                assert self.env.benchmark and (
-                    self.env.benchmark == benchmark
-                ), "Policy changed environment benchmark"
-                assert self.env.reward_space, "Policy unset environment reward space"
-                assert (
-                    self.env.reward_space.id == "IrInstructionCountOz"
-                ), "Policy changed environment reward space"
-
-                # Override walltime in the generated state.
-                state = self.env.state.copy()
-                state.walltime = timer.time
-
-                writer.write_state(state, flush=True)
-                self.states.append(state)
-
-                if not self.alive:
-                    return
-
-
-def eval_llvm_instcount_policy(policy: Policy) -> None:
+        curr_benchmarks = self.benchmarks
+        with mp.Pool(self.num_process) as pool:
+            while len(curr_benchmarks) > 0:
+                args = [ (bm, self.policy) for bm in curr_benchmarks ]
+                unsolved_benchmarks = []
+                for benchmark, state in pool.imap_unordered(eval_policy, args):
+                    if state is not None:
+                        self.q.put(state)
+                    else:
+                        # The job failed, put it to the next batch. 
+                        unsolved_benchmarks.append(benchmark)
+                curr_benchmarks = unsolved_benchmarks
```

Comment:
Here's where I would add an alternate path uses a regular `for` loop for when `num_process == 1`. If we drop the retry loop too, perhaps something like:

```suggestion
        if self.num_process > 1:
            with mp.Pool(self.num_process) as pool:
                args = [(bm, self.policy) for bm in self.benchmarks]
                for state in pool.imap_unordered(eval_policy, args):
                    self.q.put(state)
        else:
            for benchmark in self.benchmarks:
                state = eval_policy((benchmark, self.policy))
                self.q.put(state)
```